### ISSUE TITLE
Multiple dev site environment in subfolders.

### ISSUE THREAD

**Issue by klonos** [WPN-XM/WPN-XM#544]
I've been searching for an answer to this for a really long time, but no luck so far...

My server root is under ```C:\WPN-XM\www``` and under that folder I have multiple dev sites in either Drupal, WordPress or Joomla. I cannot seem to be able to enable the clean URLs feature for any Drupal website that is contained in a subfolder under my root folder - only if I place *a single* website under root of course, but that is not my intention.

So, I have for example:

`C:\WPN-XM\www\drupal7`
`C:\WPN-XM\www\drupal8`
`C:\WPN-XM\www\drupal-dev`
`C:\WPN-XM\www\joomla`
`C:\WPN-XM\www\wordpress`

...so on.

If I use ```C:\WPN-XM\www``` as a single Drupal site, clean URLs work. But I need to be simply placing more sites in subfolders under ```C:\WPN-XM\www``` and have them automatically pick up the settings for the nginx config for clean URLs. How do I do that?

**Comment by jakoch** [comment]
Hey Gregory,

if i understand correctly, then you want to set up multiple projects, each living in an own project folder inside the `www` folder and each having a vhost (its own local hostname), right?

The important part is to get the `root` and `server_name` definitions right in the nginx server block definitions. Add one definition file for each `root` + `server_name` into the `domains-enabled` folder.
Some example configs are in the `domains-disabled` folder, just copy them.
If you can't get this to work, then the rewrite rule might simply be out of date. So, from time to time these nginx configs might need an update. 

`root` is a relative path to the project, e.g. `www/<project_folder>`, but a hardcoded paths will work too, e.g. `c:\server\www\project_folder`. For instance for Wordpress: `root www/wordpress` with `server_name wordpress.dev`.

------

Here is a step by step tutorial for Drupal7.

**Drupal7**

1. https://ftp.drupal.org/files/projects/drupal-7.42.zip
2. extract into `/www` folder
      - drupal brings its own folder: `drupal-7.42`, rename it to `drupal7`
3. open in browser: http://localhost/drupal-7.42/install.php
4. select "standard"
5. enable PHP Extension "gd"
      - `php.ini`: `extension=php_gd2.dll`
6. create database with Adminer
      - http://localhost/tools/adminer/adminer.php
      - in case you need to change root pw, use `reset-db-pw.bat` in the server root
      - create table "test"
      - create new user and pw
7. use the "database", "user" and "pw"
8. configure site
9. finally Drupal is installed: http://localhost/drupal-7.42/

**Enable local VHost - http://drupal7.dev/**

1. copy the drupal7.conf from `domains-disabled` to `domains-enabled` folder
     - i've updated the config once more. you find the updated file over here: https://github.com/WPN-XM/software/blob/master/nginx/config/conf/domains-disabled/drupal7.conf
2. edit the file and check that you have:
   - `root         www/drupal7;`
   - `server_name  drupal7.dev;`
3. then add vhost with SCP
    - right-click tray, "ManageHosts"
    - click add enter `127.0.0.1` with `drupal7.dev` + "ok"
    - click main "ok"
4. restart nginx
5. finally, open http://drupal7.dev/

**Enabling Clean URLs**

1. open http://drupal7.dev/
2. login
     -  after login the URL is: http://drupal7.dev/?q=node
3. To enable clean urls go to "Configuration > Clean-Urls" or directly to 
     - http://drupal7.dev/admin/config/search/clean-urls
     - http://drupal7.dev/?q=admin/config/search/clean-urls
4. enable checkbox "enable clean urls", then submit
5. finally, open http://drupal7.dev/admin/content

In the end you get a vhost `drupal7.dev` with clean URLs, which servers from `C:\WPN-XM\www\drupal7`.

------

I will punch the other tutorials out in the next few days.

**Todo**
- [x] \www\drupal7
- [ ] \www\drupal8
- [ ] \www\drupal-dev
- [ ] \www\joomla
- [ ] \www\wordpress
- [ ] add step-by-step setup instructions to the documentation 
    - [ ] add section "PHP Applications"
       - [ ] add sections for " -App- Setup"
    - [ ] add section "Nginx" - "Configuration"       
        - document nginx.conf and domains-disabled / domains-enabled
        - server blocks
        - "Setting up local domains"

**Comment by jakoch** [comment]
**Install Instructions for Drupal8**

1. Download https://ftp.drupal.org/files/projects/drupal-8.0.4.zip
2. extract the folder `drupal-8.0.4` into the `server\www` folder
3. rename `drupal-8.0.4` folder to `drupal8`
4. browse to http://localhost/drupal8 and start installation
  - Note: don't switch the language - stay on "English". If you change the language you will get a strange "No Internet connection" error - at least i got it and i couldn't solve it. Initially i thought the PHP extensions "curl" is missing or disabled and so automatic fetching for translations from the drupal ftp server was blocked, but i can't find the reason.
5. Finally: "Congratulations, you installed Drupal!"
   - URL is http://localhost/drupal8/
      
At this point you need rewrite rules to support the Clean URLs, which are active by default in D8.
By default URLs like  http://localhost/drupal8/admin/config won't work out of the box.
You might use http://localhost/drupal8/index.php?q=admin/config instead.

**Add hosts file entry for local development domain `drupal8.dev`**

1. Server Control Panel: right-click tray, "ManageHosts"
2. click add enter 127.0.0.1 with `drupal8.dev` + "ok"
3. click main "ok"
4. restart nginx

With local host enabled it seems that rewriting for the index URL fails, because all menu links will have `/drupal8/` in them, which is wrong, e.g http://drupal8.dev/drupal8/admin/config 
But URLs like http://drupal8.dev/admin/config work.... 

**Not solved, yet. Either a misconfiguration in my drupal8.conf or  a bug in Drupal.**

**Comment by jakoch** [comment]
**Install Instructions for Joomla**

1. Download https://github.com/joomla/joomla-cms/releases/download/3.4.8/Joomla_3.4.8-Stable-Full_Package.zip
2. extract into `server\www\joomla`
3. browse: http://localhost/joomla/

4. and find out that http://localhost/joomla/installation/index.php triggers 
**Fatal error: Cannot use Joomla\String\String as String because 'String' is a special class name in D:\server-test\www\joomla\libraries\vendor\joomla\registry\src\Format\Json.php on line 12**
So Joomla 3.4.8 is not compatible with PHP7... 

5. downgrade to PHP5.6
6. Follow installation, until Joomla successfully installed 
    - URL http://localhost/joomla/
7. login
     - after login the URLs will fail http://localhost/joomla/index.php/site-settings
8. configure Nginx rewriting by adding `joomla.conf` to `server\nginx\conf\domains-enabled`
    - https://github.com/WPN-XM/software/blob/master/nginx/config/conf/domains-disabled/joomla.conf
9. create local host entry: `joomla.dev`
10 finally enjoy short urls and Joomla! 
      -  http://joomla.dev/administrator/index.php

http://joomla.dev/index.php/site-settings

**Comment by jakoch** [comment]
**Install Instructions for Wordpress**

1. Download https://wordpress.org/wordpress-4.4.2.zip
2. Extract the `wordpress` folder into `server\www\
3. Browse to http://localhost/wordpress/wp-admin/install.php
4. adjust db password @ `server\www\wordpress\wp-config.php` 
5. proceed with install
6. after install, login: http://localhost/wordpress/wp-login.php
7. create local host `wordpress.dev`
8. configure Nginx rewriting by adding `wordpress.conf` to `server\nginx\conf\domains-enabled` and restart Nginx
   - https://github.com/WPN-XM/software/blob/master/nginx/config/conf/domains-disabled/wordpress.conf
9. configure your permalinks: http://localhost/wordpress/wp-admin/options-permalink.php

**Comment by jakoch** [comment]
Everything is working fine, except the buggy Drupal8.

**Comment by klonos** [comment]
Sorry Jens for not getting back sooner. I feel ashamed that you have spent time to test and compile this together while I couldn't find some time to give feedback, but I was swamped by Real Life™.

Well, here it is now...

>if i understand correctly, then you want to set up multiple projects, each living in an own project folder inside the ```www``` folder and each having a vhost (its own local hostname), right?

Well, not exactly, my environment is set as a single hostname under ```localhost.dev``` (127.0.0.1 host entry) which works for both http as well as https.

Yes, each project will live in its own subfolder under ```C:\WPN-XM\www```, but it will not have it's own subdomain - it will have a sub-path of that local domain instead. Like so:

`C:\WPN-XM\www\drupal7` -> `http://localhost.dev/drupal7`
`C:\WPN-XM\www\drupal8` -> `http://localhost.dev/drupal8`
`C:\WPN-XM\www\drupal-dev` -> `http://localhost.dev/drupal-dev`
`C:\WPN-XM\www\joomla` -> `http://localhost.dev/joomla`
`C:\WPN-XM\www\wordpress` -> `http://localhost.dev/wordpress`

So, how I expect this to work (and how it did work when I was using another stack with apache), is that I will **not** have to create a separate vhost entry for each project each time I place a subfolder under `C:\WPN-XM\www`. I expect the clear URL feature to work out-of-the-box as I randomly create project folders.

>The important part is to get the ```root``` and ```server_name``` definitions right in the nginx server block definitions. In other words: you need multiple server blocks.

That's what I'm trying to avoid. The need to have to create vhosts/```server_name``` definitions each time.

> Some example configs are in the `domains-disabled` folder, just copy them.
If you can't get this to work, then the rewrite rule might simply be out of date. So, from time to time these nginx configs might need an update.

That could be the case, but what I would hope for here is that see how the webinterface auto-detects that project structure and auto-generates the list of links for the projects?:

![wpn-xm-webinterface-project_list](https://cloud.githubusercontent.com/assets/2423362/13368518/3b66161c-dd3f-11e5-9dc8-30833272e9f0.png)

...well how about if we had some sort of way to have either a button on the SCP and a link/button on the webinterface called "Add project". Then you'd be able to choose either a "new generic html" or "new generic php" project (things that do not need special rewrite rules) as well as a "new Drupal 7 project...", "new Joomla project..." etc. and that option would auto-create the folder under `C:\WPN-XM\www`, copy the respective .conf file under `domains-enabled` and restart nginx.

That's what I mean.

**Comment by jakoch** [comment]
Oh, i'm sorry. I got your question wrong.

It might work with a small change in the main server block and additional changes to the normal rewrite rules (take subfolder into account). But, i don't see how it could work out-of-the-box, because the rewrites are application specific... :(

    server_name  localhost localhost.dev;
    root         www;

`localhost.dev` -> `C:\WPN-XM\www`
For subfolders 
  - `localhost.dev/wordpress` -> `C:\WPN-XM\www\wordpress`

The other solution is to create .confs with server blocks and add the `server_name localhost.dev/subfolder` and adjust the rewrite rule for subfolder.
This is the cleaner solution, because the site config groups all directives 
and keeps the main block clean.

---

I really wonder how other stacks do this?
The knowledge about the rewrite rules has to come from somewhere. 
Its either a combination of .htaccess rewrite rules from the project folder with pre-configuration or some kind of boxing/prepackaging. And that brings us to stacks, which use prepackaging: 
   - XAMPP with Bitnami packaging and 
   - ZendServer with Zend packaging (zpk)

---

> how the webinterface auto-detects that project structure and auto-generates the list of links for the projects?

Basically by scanning/grabbing all existing folders in www and prefixing them with `http://localhost`.
I could 
- [ ] scan the domains-enabled folder for server_names and insert stuff into the hosts file automatically and 
- [ ] replace the URLs in the project box with "hosts" URLs instead "localhost/folder"

> auto-create the folder under C:\WPN-XM\www, copy the respective .conf file under domains-enabled and restart nginx 

That's the simple part and i like the idea. It will work for projects without rewriting.
But the "new Joomla project" is actually a fully-fledged "Install Joomla" task...

  -  button "Add Project" / "New Project" 
     -  "Install Joomla" / "app skeleton" / "app template"
         - TaskRunner needed
         - Tasks

**Comment by klonos** [comment]
> But the "new Joomla project" is actually a fully-fledged "Install Joomla" task...

I was thinking more like:

1. User extracts downloaded package in their root ```www```. Say they end up with a ```C:\WPN-XM\www\joomla``` folder.
2. They rename this folder to ```C:\WPN-XM\www\dev.mysite.net``` or whatever.
2. Either from the SCP or the webinterface they select "Add project".
3. They either type or browse for the directory/path.
4. vhost entries and nginx .conf files copied and created and nginx restarts automagically.

Thoughts?

**Comment by klonos** [comment]
...or:

1. Projects keep being auto-detected as they are currently in the webinterface and by default they are configured as simple html/php projects, **but** there's a "configure" link/button next to each project.
2. Users can click that "configure" link/button and they are presented with options (Joomla, Drupal, WordPress etc).
3. After selecting site type, respective .conf files are created and copied in the ```sites-enabled``` folder, nginx restarts.

**Comment by klonos** [comment]
Wow!! I just took a look at the screenshot you posted in the other issue about the responsiveness of the "Projects" and "Tools" sections. Looks like the basis for what I suggested above is already in place. Fantastic!!

...the cog wheel icon for configuring each project is there. Perhaps all it needs is a simple text suffix after each project link in the list. Something like this (mockup - not real screenshot):

![wpn-xm-webinterface-project_list_mockup_with_config_type_suffix](https://cloud.githubusercontent.com/assets/2423362/13371451/8c4821c0-dd7a-11e5-81cb-4a8f23f81cf1.png)

**Comment by jakoch** [comment]
Ok, i read through the posts once more to find a solution to the original issue 
"nice URLs" "served from virtual domain with subfolder":

> C:\WPN-XM\www\drupal7 -> http://localhost.dev/drupal7

As described, this needs a second, adjusted nginx configuration file.

I would like to see these configurations provided/contributed by the community.
Either by the WPN-XM community or by the community of the application (to get a better support and a tighter integration). 

---

I will try to summarize the things suggested for implementation:

(a) an automatic setup of the virtual domain based on the application being installed and 
(b) more "manual" / control over the domains when using the webinterface

I will implement (a) by pulling the data of the "software assets" repository (which includes the "configs" and will include the tasks) into the repo of the webinterface. Thereby the webinterface gets access to all application relevant data: manifest, version, config and additional stuff for the setup. 
The key here is to to trigger "setup-domain" task (if the application provides it and the config) during the "installation" task.

(b) is "nginx vhost administration" issue #2 - now open since 6 Dec 2011

What do you think?
